src/main/java/org/opensearch/knn/index/query/ResultUtil.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.DocIdSetBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+/**
+ * Utility class used for processing results
+ */
+public final class ResultUtil {
+
+    /**
+     * Reduce the results to only include the top k results across all leaf results
+     *
+     * @param perLeafResults Results from the list
+     * @param k the number of results across all leaf results to return
+     */
+    public static void reduceToTopK(List<Map<Integer, Float>> perLeafResults, int k) {
+        // Iterate over all scores to get min competitive score
+        PriorityQueue<Float> topKMinQueue = new PriorityQueue<>(k);
+        for (int i = 0; i < k; i++) {
+            topKMinQueue.add(-Float.MAX_VALUE);
+        }
+
+        for (Map<Integer, Float> perLeafResult : perLeafResults) {
+            for (Float score : perLeafResult.values()) {
+                if (topKMinQueue.peek() != null && score > topKMinQueue.peek()) {
+                    topKMinQueue.poll();
+                    topKMinQueue.add(score);
+                }
+            }
+        }
+
+        // Reduce the results based on min competitive score
+        float minScore = topKMinQueue.peek() == null ? -Float.MAX_VALUE : topKMinQueue.peek();
```

Comment:
you might have to put a loop here to keep on getting elements from priority queue till the score is -FLOAT.MAX_VALUE

Reply:
I think this looks good. We don't need a loop here. I was thinking of some other case.

But we can do this, if minScore == Float.MAX we can return directly perLeafResults and we don't need to iterate over perLeafResults.

Reply:
sure, I added a count just because if minScore == -Float.MAX check wont be true if there are exactly k results